### DISCUSSION TITLE
Idea: Project creation using templates

### DISCUSSION THREAD

**Discussion by Fasuh** [leoafarias/sidekick#77] [Ideas]
Not really sure how to fully complete this idea template but there has to be someone who will do it first. hope it is sufficient 

# Problem Space
## Problem Statement
I'd like to be able to create a new Flutter project using Sidekick, and also to be able to select Flutter version through FVM during the creation. Additionally it could be useful to implement some way of custom template creation.
 
## Impact of the problem
I think it would greatly improve user experience with the Sidekick project, as for now you are required to use some kind of external tool to create a new Flutter project, only after that you are able to change its FVM version using Sidekick tool. It would be more optimal to use just one tool for management of your projects.

Additionally there is a whole almost-empty place for template generation, there are useful tools but they are mostly console-only, integrating them into the tool could help many users start their projects right away without, again, using external tools.
 
## How do we judge success?
I think there are two layers of the problem,
first - the ability to generate a clean Flutter project using Sidekick tool, possibly with the FVM settings, that shouldnt be that much of a problem, the most problematic element here is possibility of multiple official Flutter templates, with introduction of sound null-safety it might create a need for the tool to use different source depending on the Flutter version from FVM.

Second layer would be custom template generation, there are tools specialized with that, and i think just adding and removing custom templates could be easily done, however, i have seen voices saying if we could implement discoverable community templates, that would require a lot more work which i dont relaly think is necessary, considering the fact that i dont think many people would look for a template, rather, they would try to provide their own.
 
## Possible Solutions
List your high-level ideas for possible solutions. (Can be filled out later)
 
# Validation
## What we already know?
What data or insights do you have to validate this?

Nothing really
 
## What do we need to answer?
What assumptions are we making that need to be validated/refuted? What questions will increase our confidence in the decisions we need to make? What are the gaps in our understanding?

We are assuming that there is a way to get newest/older versions of Flutter project templates in order to be able to use them with FVM, or at least get only some of the latest ones

We are assuming that there is a way to implement an external tool like Mason (https://pub.dev/packages/mason) or make our own to provide user with a custom templates

We are assuming there is a way of storing and providing other users with community-made templates for their projects 
 
# Ready to make it
## What are we doing?
What will the project deliver? For example, an elevator pitch or read me.

The project will deliver a complete tool to kick-start your Flutter project without additional tools required.

## Why will a user want this?
It's good to get Sidekick user's feedback on this.

It seemed like my issue got a good feedback (https://github.com/leoafarias/sidekick/issues/71#issue-889789637)
 
## Visualize the solution
It's good to get Design input here: whiteboard sketches, journey map, end to end demo

## Scale and scope
It's good to get Engineering input here: t-shirt size estimate

**Comment by leoafarias** [comment]
@Fasuh Thank you for taking the time, this is extremely helpful for us to understand the scope of the problem and some solutions

**Comment by aguilaair** [comment]
Regarding community contributions, we can do something like an "official/verified" repo where people can merge their templates and then sidekick simple uses the data from that repo. Then there can be 3rd party repos, where a user can enter the repo on sidekick to include it in their library.

**Comment by leoafarias** [comment]
Want to reference something else for later. I think we should consider using `stagehand` for the template generation. Seems like a very simple setup and it's a standard way to install dart boilerplates. https://github.com/dart-lang/stagehand/tree/master/templates

**Reply by aguilaair** [reply]
Looks good!

**Reply by Fasuh** [reply]
well, main point of the idea was to allow people to define their own templates, as long as that is fulfilled i think its good

**Reply by aguilaair** [reply]
Update on this. The dart SDK now has this functionality built in: https://dart.dev/tools/dart-create
